File: cpp/93_arithmetic_expressions.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <stack>
#include <set>

int evaluate(std::vector<int> digits, std::vector<std::string> ops, std::string group){
    // stack containing intermediate results
    std::stack<double> s;
    int d = 0, opIdx = 0;
    for (char c : group){
        // apply operation to top two elements in stack
        if (c == '.'){
            double num2 = s.top();
            s.pop();
            double num1 = s.top();
            s.pop();
            // exit if division by zero
            std::string op = ops[opIdx++];
            if (op == "/" && num2 == 0){
                return 0;
            }
            // push evaluated intermediate expression to stack
            if (op == "/") s.push(num1 / num2);
            else if (op == "*") s.push(num1 * num2);
            else if (op == "+") s.push(num1 + num2);
            else s.push(num1 - num2);
        }
        else{ // push operand to stack
            s.push(double(digits[d++]));
        }
    }
    // return result if it is a positive integer
    if (s.top() > 0 && s.top() == int(s.top())) return s.top();
    // return dummy if result is not positive or not an integer
    return 0;
}

int get_streak(std::vector<int> digits){
    std::vector<std::string> ops = {"+", "-", "/", "*"};
    // five possible groupings of three binary operators
    // a b c d x y z = a z (b y (c x d))
    // a b c x d y z = a z ((b x c) x d)
    // a b c x y d z = (a x (b x c)) x d
    // a b x c d y z = (a x b) x (c x d)
    // a b x c y d z = (a x b) x c x d
    std::vector<std::string> groups = {
        "xxxx...", "xxx.x..", "xxx..x.", "xx.xx..", "xx.x.x."
    };
    std::set<int> resultSet;
    
    // calculate possible results for each permutation of digits
    // with a given list of operations and grouping of parentheses
    do{
        for (int x = 0; x < 4; x++){
            for (int y = 0; y < 4; y++){
                for (int z = 0; z < 4; z++){
                    std::vector<std::string> operations = {ops[x], ops[y], ops[z]};
                    for (std::string group : groups){
                        int result = evaluate(digits, operations, group);
                        // insert positive integer results into result set
                        if (result) resultSet.insert(result);
                    }
                }
            }
        }
    }
    while (std::next_permutation(digits.begin(), digits.end())); // permute digits after each loop
    
    // copy result set into vector and sort vector
    std::vector<int> results(resultSet.begin(), resultSet.end());
    std::sort(results.begin(), results.end());
    
    // get length of streak of consecutive positive integers starting from 1
    for (int i = 0; i < results.size(); i++){
        if (results[i] != i + 1) return i;
    }
    return -1;
}

std::string arithmetic_expressions(){
    std::string digits = "";
    int max_streak = 0;
    // generate tuple of digits (a, b, c, d) such that a < b < c < d
    for (int a = 0; a < 10; a++){
        for (int b = a + 1; b < 10; b++){
            for (int c = b + 1; c < 10; c++){
                for (int d = c + 1; d < 10; d++){
                    int streak = get_streak({a, b, c, d});
                    if (streak > max_streak){
                        max_streak = streak;
                        digits = std::to_string(1000 * a + 100 * b + 10 * c + d);
                        // handle string edge case
                        if (a == 0){
                            digits = "0" + digits;
                        }
                    }
                }
            }
        }
    }
    return digits;
}

int main(){
    std::cout << arithmetic_expressions() << "\n";
}

File: java/_93_ArithmeticExpressions.java
import java.util.*;

public class _93_ArithmeticExpressions{
    
    // generate combinations of 4 digits from 0-9
    static List<List<Integer>> digits = generateCombs(9, 4);
    
    // generate groups of 3 operators
    static List<String> ops = generateOps();
    
    // list out postfix groupings of digits and operators
    static String[] groups = new String[]{
        "xxxx...", "xxx.x..", "xxx..x.", "xx.xx..", "xx.x.x."
    };
    
    public static List<String> generateOps(){
        List<String> res = new ArrayList<>();
        String operators = "+-*/";
        // operators may appear more than once or not at all
        // sampling operators with replacement
        for (int i = 0; i < 4; i++){
            for (int j = 0; j < 4; j++){
                for (int k = 0; k < 4; k++){
                    StringBuilder newOp = new StringBuilder();
                    newOp.append(operators.charAt(i));
                    newOp.append(operators.charAt(j));
                    newOp.append(operators.charAt(k));
                    res.add(newOp.toString());
                }
            }
        }
        return res;
    }
    
    public static List<List<Integer>> generateCombs(int n, int k){
        List<List<Integer>> combs = new ArrayList<>();
        backtrackCombs(combs, new ArrayList<>(), n, k, 0);
        return combs;
    }
    
    private static void backtrackCombs(List<List<Integer>> combs, List<Integer> curr, int n, int k, int idx){
        if (curr.size() == k){
            combs.add(new ArrayList<>(curr));
            return;
        }
        for (int i = idx; i <= n; i++){
            curr.add(i);
            backtrackCombs(combs, curr, n, k, i + 1);
            // avoiding ambiguity of value vs. position
            // when removing an element from a list of integers
            // specifying an Integer object removes a value
            curr.remove(Integer.valueOf(i));
        }
    }
    
    public static List<List<Integer>> generatePerms(List<Integer> nums){
        List<List<Integer>> perms = new ArrayList<>();
        backtrackPerms(perms, new ArrayList<>(), nums, new boolean[nums.size()]);
        return perms;
    }
    
    private static void backtrackPerms(List<List<Integer>> perms, List<Integer> curr, List<Integer> nums, boolean[] used){
        if (curr.size() == used.length){
            perms.add(new ArrayList<Integer>(curr));
            return;
        }
        for (int i = 0; i < used.length; i++){
            if (!used[i]){
                used[i] = true;
                curr.add(nums.get(i));
                backtrackPerms(perms, curr, nums, used);
                curr.remove(nums.get(i));
                used[i] = false;
            }
        }
    }
    
    private static int countResults(List<Integer> digits){
        List<List<Integer>> perms = generatePerms(digits);
        Set<Integer> results = new HashSet<>();
        
        for (List<Integer> perm : perms){
            for (String op : ops){
                for (String group : groups){
                    int res = evaluateExpression(perm, op, group);
                    if (res > 0){
                        results.add(res);
                    }
                }
            }
        }
        // calculate consecutive streak of integer results starting from 1
        int i = 0;
        while (results.contains(i + 1)) i++;
        return i;
    }
    
    // evaluate postfix expression given digits and operators
    private static int evaluateExpression(List<Integer> perm, String op, String group){
        Stack<Double> results = new Stack<>();
        int numIdx = 0, opIdx = 0;
        for (char c : group.toCharArray()){
            if (c != 'x'){
                char currOp = op.charAt(opIdx++);
                if (currOp == '+'){
                    results.push((double) results.pop() + results.pop());
                }
                else if (currOp == '*'){
                    results.push(results.pop() * results.pop());
                }
                else if (currOp == '-'){
                    double op2 = results.pop();
                    double op1 = results.pop();
                    results.push(op1 - op2);
                }
                else{
                    double op2 = results.pop();
                    if (op2 == 0) return 0;
                    double op1 = results.pop();
                    results.push(op1 / op2);
                }
            }
            else{
                results.push((double) perm.get(numIdx++));
            }
        }
        double res = results.pop();
        if (res < 0 || (int) res != res) return 0;
        return (int) res;
    }
    
    public static void main(String[] args){
        int maxStreak = 0;
        String res = "";
        
        for (List<Integer> comb : digits){
            int streak = countResults(comb);
            if (streak > maxStreak){
                maxStreak = streak;
                StringBuilder sb = new StringBuilder();
                for (int d : comb) sb.append(d);
                res = sb.toString();
            }
        }
        
        System.out.println("Digits that result in longest set of consecutive integers using arithmetic operations: " + res); // 1258
    }
}
